lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -620,6 +620,93 @@ class swift::ParseableInterfaceBuilder {
 
 namespace {
 
+/// Keeps track of the various reasons the module interface loader needed to
+/// fall back and rebuild a module from its interface.
+struct ModuleRebuildInfo {
+  enum class ModuleKind {
+    Normal,
+    Cached,
+    Forwarding,
+    Prebuilt
+  };
+  struct OutOfDateModule {
+    Optional<serialization::Status> serializationStatus;
+    ModuleKind kind;
+    SmallVector<std::string, 10> outOfDateDependencies;
+  };
+  llvm::StringMap<OutOfDateModule> outOfDateModules;
+
+  /// Sets the kind of a module that failed to load.
+  void setModuleKind(StringRef path, ModuleKind kind) {
+    outOfDateModules[path].kind = kind;
+  }
+
+  /// Sets the serialization status of the module at \c path. If this is
+  /// anything other than \c Valid, a note will be added stating why the module
+  /// was invalid.
+  void setSerializationStatus(StringRef path, serialization::Status status) {
+    outOfDateModules[path].serializationStatus = status;
+  }
+
+  /// Registers an out-of-date dependency at \c depPath for the module
+  /// at \c modulePath.
+  void addOutOfDateDependency(StringRef modulePath, StringRef depPath) {
+    outOfDateModules[modulePath].outOfDateDependencies.push_back(depPath);
+  }
+
+  Optional<const char *> invalidModuleReason(serialization::Status status) {
+    using namespace serialization;
+    switch (status) {
+      case Status::FormatTooOld:
+        return "compiled with an older version of the compiler";
+      case Status::FormatTooNew:
+        return "compiled with a newer version of the compiler";
+      case Status::Malformed:
+        return "malformed";
+      case Status::TargetIncompatible:
+        return "compiled for a different target platform";
+      case Status::TargetTooNew:
+        return "target platform newer than current platform";
+      default: return None;
```

Comment:
Nitpick: if you're returning StringRef `None` makes sense here, but with `char *` you already have a null state.